### PR TITLE
backport testing and config changes

### PR DESCRIPTION
**Pull request by Syxton**

### CODE REVIEW

File: block_massaction.php [hunk 1]
```diff
@@ -90,7 +94,7 @@ public function get_content(): stdClass {
         if ($this->page->user_is_editing()) {
 
             $applicableformatkey = 'course-view-' . $COURSE->format;
-            $iscoursecompatible = in_array($applicableformatkey, $this->applicable_formats())
+            $iscoursecompatible = in_array($applicableformatkey, array_keys($this->applicable_formats()))
```
Comment (PhMemmel) [inline block_massaction.php:97]: looks like a rebase to current `39_311_stable` would clean this up.

Comment (Syxton) [reply block_massaction.php]: thanks!

File: settings.php [hunk 2]
```diff
@@ -37,5 +37,26 @@
...
+        foreach (array_keys($pluginmanager->get_installed_plugins('format')) as $plugin) {
+            $plugins[$plugin] = new lang_string('pluginname', 'format_' . $plugin);
+        }
+        // Sort alphabetically. Custom sort function needed, because lang_string is an object.
+        uasort($plugins, function($a, $b) {
+            return strcmp($a->out(), $b->out());
+        });
+
+        // These are the formats supported by the maintainer.
+        $supportedformatsbydefault = ['weeks' => 1, 'topics' => 1, 'topcoll' => 1];
```
Comment (PhMemmel) [inline settings.php:52]: We should add the currently supported plugins in 3.9 to 3.11, right?
https://github.com/Syxton/moodle-block_massaction/blob/a3c388a722672c9ca494ad31354b73495ddccec7/block_massaction.php#L51-L53

**Comment by Syxton** [comment]
@PhMemmel Yeah, so if we were to add behat tests for a course format, that would come with a version update and a change to the default supported formats (using those functions I made).  When the behat runs, it would run with those defaults and the format would be checked.

**Comment by PhMemmel** [comment]
Alright.

When trying to upgrade I see this:
grafik

The apparently missing formats seem to be "format_gps" und "format_scorm" in my instance which do not seem to be course format plugins. This makes me think that we do not only catch course formats here, but also accidentally list other plugins. This probably is also a problem of our moodle 4.0 branch. Is there a better way to determine only(!) course format plugins?

File: tests/behat/topcoll.feature [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      | assign   | TC2    | 5        | Test Activity5 | Test page description | 4       |
+    When I log in as "teacher1"
+    And I am on "Test course" course homepage with editing mode on
+    And I add the "Mass Actions" block
+    # Everything is setup now, let's do the real test.
+    And I click on "Open all" "text"
+    And I click on "Test Activity2 Checkbox" "checkbox"
+    And I click on "Test Activity5 Checkbox" "checkbox"
+    And I click on "Indent (move right)" "link" in the "Mass Actions" "block"
+    Then "#section-1 li.modtype_page div.indent-1" "css_element" should exist
```
Comment (PhMemmel) [inline tests/behat/topcoll.feature:70]:
```suggestion
    Then "#section-1 li.page div.mod-indent-1" "css_element" should exist
```

**Comment by PhMemmel** [comment]
One more though: As we do not want to put any effort into the 39-311 branch anymore, we also could leave the tests as they are and do not port the refactoring of the tests we mainly did, because course formats can for example decide to support indent/outdent in moodle 4.0.

We have tested 39-311 with the configured standard course formats and behat tests are running fine. We are not interested in adding any more course formats as defaults either, so maybe we just remove the test rework from this PR and only backport the ability to configure the formats?

**Comment by Syxton** [comment]
format_gps is a course format, scorm isn't as far as I can tell.
https://moodle.org/plugins/format_gps

**Comment by Syxton** [comment]
@PhMemmel Another way is to use $formatplugins = core_plugin_manager::instance()->get_plugins_of_type('format');

I'm really not sure I understand the difference between get_plugins_of_type and get_installed_plugins

**Comment by PhMemmel** [comment]
Just tested `core_plugin_manager::instance()->get_plugins_of_type('format')` in my moodle 311 dev instance, this seem to resolve the issue. Both `format_gps` and `format_scorm` are gone in the settings menu.

BTW: I doublechecked: There is no scorm course format installed in my instance. This must be something else which is found by the function.

But looks like using `core_plugin_manager::instance()->get_plugins_of_type('format')` solves this issue anyway. I will send a PR for 4.0 branch with this change. Can you push it to this PR?

**Comment by PhMemmel** [comment]
In my 3.11.6+ dev environment:

I reverted `block_massaction` to current `39_311_stable`, uninstalled completely, reinstalled. Then upgraded to the branch of this PR. Defaults are now being properly selected. No issues with listed but not really existing formats anymore.

I tested if I can deactive and activate the block for a specific course format, seems to work.

Guess that's good to go now :+1:

**Pull request merged**